src/inet/routing/leach/Leach.cc

```diff
@@ -0,0 +1,450 @@
+        TDMADelayCounter = 1;
+        event = new cMessage("event");
+        roundDuration = dblrand(0) * 10;
+        wasCH = false;
+
+        vector<nodeMemoryObject> nodeMemory(numNodes); // Localized NCH node memory with CH data
+        vector<TDMAScheduleEntry> nodeCHMemory(numNodes); // Localized CH memory
+        vector<TDMAScheduleEntry> extractedTDMASchedule(numNodes); // Localized NCH node memory with extracted TDMA data
+
+    } else if (stage == INITSTAGE_ROUTING_PROTOCOLS) {
+        registerService(Protocol::manet, nullptr, gate("ipIn"));
+        registerProtocol(Protocol::manet, gate("ipOut"), nullptr);
+    }
+}
+
+void Leach::start() {
+    // Search the 802154 interface
+    configureInterfaces();
+
+    // schedules a random periodic event
+    event->setKind(SELF);
+    scheduleAt(simTime() + uniform(0.0, par("maxVariance").doubleValue()),
+            event);
+}
+
+void Leach::stop() {
+    nodeMemory.clear();
+    nodeCHMemory.clear();
+    extractedTDMASchedule.clear();
+    TDMADelayCounter = 1;
+}
+
+void Leach::configureInterfaces() {
+    int numInterfaces = interfaceTable->getNumInterfaces();
+    for (int i = 0; i < numInterfaces; i++) {
+        NetworkInterface *networkInterface = interfaceTable->getInterface(i);
+        // Only single wireless interface is required
+        if (networkInterface->isWireless()) {
+            wirelessInterface = networkInterface;
+            break;
+        }
+    }
+}
+
+void Leach::handleMessageWhenUp(cMessage *msg) {
+    // if node is sending message
+    if (msg->isSelfMessage()) {
+        double randNo = dblrand(1);
+        double threshold = generateThresholdValue(round);
+
+        if (randNo < threshold && wasCH == false) {
+            setLeachState(ch);               // Set state for GUI visualization
+            wasCH = true;
+            handleSelfMessage(msg);
+        }
+
+        round += 1;
+        int intervalLength = 1.0 / clusterHeadPercentage;
+        if (fmod(round, intervalLength) == 0) { // reset values at end of number of subintervals
+            wasCH = false;
+            nodeMemory.clear();
+            nodeCHMemory.clear();
+            extractedTDMASchedule.clear();
+            TDMADelayCounter = 1;
+        }
+
+        // schedule another self message every time new one is received by node
+        event->setKind(SELF);
+        scheduleAt(simTime() + roundDuration, event);
+        // if node is receiving message
+    } else if (check_and_cast<Packet*>(msg)->getTag<PacketProtocolTag>()->getProtocol()
+            == &Protocol::manet) {
+        processMessage(msg);
+    } else {
+        throw cRuntimeError("Message not supported %s", msg->getName());
+    }
+}
+
+void Leach::handleMessageWhenDown(cMessage *msg) {
+    delete msg;
+    OperationalBase::handleMessageWhenDown(msg);
+}
+
+void Leach::handleSelfMessage(cMessage *msg) {
+    if (msg == event) {
+        if (event->getKind() == SELF) {
```

Comment:
I don't see the SELF kind necessary. you already know that this is a self message as `msg == event`. Also there is an explicit `isSelfMessage()` on the messages which gives exactly this info if needed.

Reply:
Noted, sorry about that. Seems to be a remnant from when I was testing with multiple self message types.
Resolved in https://github.com/inet-framework/inet/pull/929/commits/2a7da9283b9ece61cc0a94d03fb3902cc4ec35d0